zipline/data/data_portal.py

```diff
@@ -311,6 +323,146 @@ def _get_daily_data(self, asset_int, column, dt):
         else:
             return raw_value
 
+    def _get_history_daily_window(self, sids, end_dt, bar_count, field_to_use):
+        """
+        Internal method that returns a dataframe containing history bars
+        of daily frequency for the given sids.
+        """
+        data = []
+
+        day = end_dt.date()
+        day_idx = tradingcalendar.trading_days.searchsorted(day)
+        days_for_window = tradingcalendar.trading_days[
+            (day_idx - bar_count + 1):(day_idx + 1)]
+
+        ends_at_midnight = end_dt.hour == 0 and end_dt.minute == 0
+
+        if len(sids) == 0:
+            return pd.DataFrame(None,
+                                index=days_for_window,
+                                columns=None)
+
+        for sid in sids:
+            sid = int(sid)
+
+            # get the start and end dates for this sid
+            if sid not in self.asset_start_dates:
+                asset = self.asset_finder.retrieve_asset(sid)
+                self.asset_start_dates[sid] = asset.start_date
+                self.asset_end_dates[sid] = asset.end_date
+
+            if ends_at_midnight or \
```

Comment:
if the history window ends at midnight UTC, and we're in daily mode, we can use daily data for the entire window (instead of building the last value from the last day's minute data).